Sources/SwiftParser/Types.swift

```diff
@@ -194,15 +194,21 @@ extension Parser {
       return RawTypeSyntax(RawMissingTypeSyntax(arena: self.arena))
     }
 
-    // '.Type', '.Protocol', '?', '!', and '[]' still leave us with type-simple.
     var loopCondition = LoopProgressCondition()
     while loopCondition.evaluate(currentToken) {
-      if !stopAtFirstPeriod, let (period, type) = self.consume(
-        if: { [.period, .prefixPeriod].contains($0.tokenKind) },
-        followedBy: { $0.isContextualKeyword(["Type", "Protocol"])}
-      ) {
-        base = RawTypeSyntax(RawMetatypeTypeSyntax(
-          baseType: base, period: period, typeOrProtocol: type, arena: self.arena))
+      if !stopAtFirstPeriod, self.at(.period) || self.at(.prefixPeriod) {
+        let period = self.consumeAnyToken()
+        if self.atContextualKeyword("Type") || self.atContextualKeyword("Protocol") {
+          let typeOrProtocol = self.consumeAnyToken(remapping: .contextualKeyword)
+          base = RawTypeSyntax(RawMetatypeTypeSyntax(
+            baseType: base, period: period, typeOrProtocol: typeOrProtocol,
+            arena: self.arena))
+        } else {
+          let (name, generics) = self.parseTypeNameWithGenerics()
+          base =  RawTypeSyntax(RawMemberTypeIdentifierSyntax(
+            baseType: base, period: period, name: name,
```

Comment:
I'll have a look

Reply:
This wasn't working because we actually parse `.` as an unknown token. I'll fix that up and add a couple diagnostics to cover the `Foo. member` and `Foo.Type.Bar` cases since the grammar doesn't allow either of those.